Update linalg.py

JayBhatt05 opened the pull request:
Moved up the matrix_rank() method above the inv() method to check singularity of a matrix in inv() method definition. If matrix is singular, an exception is thrown. Solves incorrect matrix inversion of a singular matrix.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

JayBhatt05 committed:
Update linalg.py

Moved up the matrix_rank() method above the inv() method to check singularity of a matrix in inv() method definition. If matrix is singular, an exception is thrown

rkern commented:
Can you revert the change in location of the `matrix_rank()` definition? It does not need to appear in the file before other functions that call it.

JayBhatt05 commented:
Understood, will do it now

On Wed, 13 Sep 2023 at 9:02 PM, Robert Kern ***@***.***>
wrote:

> Can you revert the change in location of the matrix_rank() definition? It
> does not need to appear in the file before other functions that call it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/numpy/numpy/pull/24692#issuecomment-1717865174>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBE3T5DQ5V23ZLBT4TAAIJTX2HGYRANCNFSM6AAAAAA4WQRFGQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

JayBhatt05 committed:
Update linalg.py

Resolved incorrect inversion of singular matrix in lumpy.linalg.inv() method by checking whether rank of a is equal to number of rows in a using matrix_rank() method

JayBhatt05 committed:
Update linalg.py

Resolved incorrect inversion of singular matrix in numpy.linalg.inv() method by checking whether rank of a is equal to number of rows in a using matrix_rank() method.

eric-wieser commented on numpy/linalg/linalg.py:
```diff
@@ -543,6 +543,8 @@ def inv(a):
     with errstate(call=_raise_linalgerror_singular, invalid='call',
```
Isn't `_raise_linalgerror_singular` supposed to detect this already? I worry that this is going to make matrix inversion much more expensive.

JayBhatt05 replied:
I executed a sample code to check the inv() function and it is resulting in an incorrect output
![Test](https://github.com/numpy/numpy/assets/139049460/d01c510b-d77c-45c0-89a8-f1cd4a217185)

rkern replied:
`_raise_linalgerror_singular` isn't doing any checking; it's just transforming the generic exception from the `gufunc` machinery (and in turn from the underlying LAPACK routine) into the informative `LinAlgError`.

rkern commented:
C.f. https://github.com/numpy/numpy/issues/23793#issuecomment-1557951725
